proposals/mpi-operator-proposal.md

```diff
@@ -0,0 +1,224 @@
+## Motivation
+Kubeflow currently supports distributed training of TensorFlow models using
+[tf-operator](https://github.com/kubeflow/tf-operator), which relies on
+centralized parameter servers for coordination between workers. An alternative
+is a decentralized approach in which workers communicate with each other
+directly via the MPI `allreduce` primitive, without using parameter servers.
+NVIDIA's [NCCL](https://developer.nvidia.com/nccl) library has implemented most
+of the MPI primitives efficiently on GPUs, and Uber's
+[Horovod](https://github.com/uber/horovod) makes it simple to do multi-gpu and
+multi-node training using TensorFlow; the TensorFlow team is also working on an
+`allreduce`
+[implementation](https://github.com/tensorflow/tensorflow/tree/master/tensorflow/contrib/distribute).
+Compared to parameter servers, this approach arguably is more
+[bandwidth optimal](http://www.cs.fsu.edu/~xyuan/paper/09jpdc.pdf) and
+[scales better](https://eng.uber.com/horovod/).
+
+However, right now there is no easy way to launch this type of training jobs on
+Kubernetes. By providing a CRD and a custom controller, we can make
+allreduce-style distributed training as simple as training on a single node.
+
+## Goals
+* Provide a common Custom Resource Definition (CRD) for defining a single-gpu,
```

Comment:
I may be missing something, but what is the value of a single-gpu training job using all-reduce? I'm not even sure nccl would run in that case because forming a singleton ring is rather odd.

Reply:
It's important to support a single-GPU use case as well to minimize friction between doing a single-GPU and multi-GPU training.  In case of a single-GPU Horovod optimizes away allreduce and so there's no performance penalty.

Reply:
Yes exactly. You can do your debugging and development on a single-gpu job. When you're ready to train the model for real, just change a single number to scale it up.